### PR TITLE
Add model nicknames

### PR DESCRIPTION
**Pull request by FynnBe**
This PR adds nicknames for all (external/zenodo) models. They are added to the resource.yaml files (in the future in each PR; we should close all current PRs upon merging this one).

The nicknames are an adjective from [adjectives.txt](https://github.com/FynnBe/collection-bioimage-io/blob/main/adjectives.txt ) and an animal name from [animals.yaml](https://github.com/FynnBe/collection-bioimage-io/blob/main/animals.yaml), which also holds an icon, which is saved as `nickname_icon`.

The fields `nickname` and 'nickname_icon` are saved in every (external model) rdf.yaml under config:bioimageio and also included in the [collection summary](https://github.com/FynnBe/collection-bioimage-io/blob/gh-pages/collection.json)

The `nickname_icon` is a unicode character. I know this might not be suitable directly for the website, so we might want to add a mapping of those unicode icons to image (png/svg) icons...

### CODE REVIEW

File: collection/10.5281/zenodo.6028097/resource.yaml [hunk 1]
```diff
@@ -12,3 +12,5 @@ doi: 10.5281/zenodo.6028097
 type: model
 owners:
 - 77626
+nickname: organized badger
```
Comment (oeway) [inline collection/10.5281/zenodo.6028097/resource.yaml:15]: Nice!

Comment (FynnBe) [reply collection/10.5281/zenodo.6028097/resource.yaml:15]: my favorite is the 'content microbe' followed by 'rude pig'^^
I swear, these are purely random!

**Review by oeway — COMMENTED** [review]
Nice!

I think we can change `nickname_icon` to `icon`, since we already have the icon field, I don't see why we need both nickname_icon and icon.

And for the nickname, maybe use dash to connect the adjective and the animal, so it will be easier to read when we place it in an URL.

This also lead to the question that, whether we should abandon the doi-based id, and only use the nick name as the id. Maybe not since we cannot guarantee the uniqueness for the nickname. But then I am confused a bit on what's the actual usage of the nick name. We can display it for sure and make it fun, we can also make the nickname searchable, is that all the purpose of it? 

And also can users define their model nickname in the model yaml? Maybe place it under the config.bioimageio field? And if the nickname is defined both in the yaml and the resource.yaml, which one has priority?

File: collection/10.5281/zenodo.6028097/resource.yaml (hunk 1 above)
Comment (oeway) [reply collection/10.5281/zenodo.6028097/resource.yaml:15]: That's so cute, and cool!

I am now also wondering whether a certain combination of nickname can be offensive or inappropriate to certain cultural background.

**Comment by FynnBe** [comment]
> I think we can change nickname_icon to icon, since we already have the icon field, I don't see why we need both nickname_icon and icon.

I did not want to overwrite the `icon` field... as I am not checking if it exists and always write them... 

> And for the nickname, maybe use dash to connect the adjective and the animal, so it will be easier to read when we place it in an URL.

I see. Could we use an underscore and just treat it as a space? That would not collide with some of the adjectives, e.g. 'self-absorbed snail'. I would leave the space in the nickname for prettier display, but treat is equivalent if it is an underscore (I'll just replace all underscores with spaces for the lookup...). So we an copy the id with underscore, why displaying with space?
No strong opinion on this though...

> Maybe not since we cannot guarantee the uniqueness for the nickname. 

I would keep the doi as it is very convenient to retrieve the original record (and a bit more proper). But we can guarantee uniqueness for the nicknames as well.. I check on generation, but we should also add a test on the final repo as they might be changed manually... yeah, I'll add a check.

> And also can users define their model nickname in the model yaml? Maybe place it under the config.bioimageio field? And if the nickname is defined both in the yaml and the resource.yaml, which one has priority?

currently I just overwrite it if it exists. We can check for a given nickname though.... with the above mentioned uniqueness check the PR should then fail (the uniqueness problem had me overwrite it in the first implementation)

**Comment by constantinpape** [comment]
This looks good overall, a few thoughts from my side:

> whether we should abandon the doi-based id, and only use the nick name as the id.

I wouldn't do it, since direct linking to the DOI and the corresponding zenodo entry is very helpful for development.

> But then I am confused a bit on what's the actual usage of the nick name. We can display it for sure and make it fun, we can also make the nickname searchable, is that all the purpose of it?

I think the idea would be to display it on the model card as well as `nick name` or `alias`. And then make the models also searchable via this and enable download via the nickname in `bioiageio.core/spec` (if that's feasible).

File: collection/10.5281/zenodo.6028097/resource.yaml (hunk 1 above)
Comment (FynnBe) [reply collection/10.5281/zenodo.6028097/resource.yaml:15]: yes, I have tried to pick mostly positive and a few relatively benign negative adjectives. I think `rude pig` is as offensive as it gets and I hope everyone is fine with that. If not we can modify `adjectives.txt` and `animals.yaml` accordingly. Also for externally set nicknames I will enforce that they are a combination that could have been generated randomly and that the nickname_icon (if present) matches the animal. If no nickname_icon is present it is added according to the mapping in `animals.yaml`

**Comment by FynnBe** [comment]
>  [...] download via the nickname in bioiageio.core/spec (if that's feasible).

yes, exactly, I'll add this `load_resource_description('rude pig')` is coming!

**Comment by FynnBe** [comment]
note that in this implementation the nickname is a nickname for the resource id and is not version sensitive

**Comment by constantinpape** [comment]
> note that in this implementation the nickname is a nickname for the resource id and is not version sensitive

Yes, I think this is how it should be. For more fine-grained control we should always resort to `resource_id/version_id`.

**Comment by FynnBe** [comment]
> According to what you described, the nickname is more like an alternative id, especially when you will also enforce uniqueness, and for that I would try to use dash instead of space, sometimes it's can be confusing, say, if one inserted two spaces in between. I won't worry about the display, we can always transform it by converting dash to space on the card (e.g. display self-absorbed-snail as Self-absorbed Snail). but it's more important to make sure we have a safer rule on the nickname itself. It's a bit unnecessary to check all types combinations (dash or space) everywhere.

yes, good point, but then underscore '_' is better suited I think, so we can easily map from 'self-absorbed_snail' to 'Self-absorbed Snail', i.e. we can forbid our underscore separator to be present in the adjective and animal name.

**Comment by FynnBe** [comment]
I now removed
- rude
- arrogant
- self-absorbed
- dirty
- aggressive
- hateful
- hyperactive
- silly
- slow
- smelly  (it's a shame, I was really looking forward for 'smelly cat')
- stubborn

potentially negative adjectives that are still listed:
- angry
- greedy 
- demanding
- insecure
- moody
- naked
- poisonous
- venomous
- rebellious
- wild

to take a more detailed look check out https://github.com/FynnBe/collection-bioimage-io/blob/main/adjectives.txt

**Comment by oeway** [comment]
> yes, good point, but then underscore '_' is better suited I think, so we can easily map from 'self-absorbed_snail' to 'Self-absorbed Snail', i.e. we can forbid our underscore separator to be present in the adjective and animal name.

I am thinking maybe we should just display the id as is `self-absorbed-snail`. Do you see big difference bewteen `self-absorbed snail` and `self-absorbed-snail`? It can also cause confusion, since if we display as is, user can also just copy it and use it as an id.  Very much similar to github repo names, they don't change it when displaying it, like that user can also copy it.

In general, hyphen is often preferred when used as id. It's more common to have hyphen in URL, github repo name, html element id etc. Mostly because it saves keystroke when typing compared to underscore.

**Comment by FynnBe** [comment]
I see the issue, but there is no way of splitting the `self-absorbed-t-rex` correctly, whereas the `self-absorbed_t-rex` is not a problem.
I suggest we use the underscore and also display it with that, so it is directly copyable. For matching in bioimageio.spec and bioimage.core we could map all to '-' to be able to always resolve it... what do you think @oeway ?

**Comment by FynnBe** [comment]
> I see the issue, but there is no way of splitting the `self-absorbed-t-rex` correctly, whereas the `self-absorbed_t-rex` is not a problem

On the other hand we can just have a little helper for splitting it up...  and outside of the collection repository splitting the nickname should not be used anyway. the 'only-dash-solution' is probably the easiest for consumers to deal with...

**Comment by oeway** [comment]
> the 'only-dash-solution' is probably the easiest for consumers to deal with...

It looks nicer and easier to type too!

**Comment by oeway** [comment]
Here is another thought, each model has a dedicated `name` and the previous discuss with Anna is that we will display both nickname and the previous name on the card. So we can generate the display name already when we generate a nickname or take it from the user, and stored in the RDF. In other words, we don't really need to split the nickname for nicer display if that make really a difference.

On the other hand, I think it's fine to just display `self-absorbed-t-rex` everywhere, it won't make a big different compared to `self-absorbed t-rex`.

I will go for the easy way.

**Comment by FynnBe** [comment]
> Here is another thought, each model has a dedicated name and the previous discuss with Anna is that we will display both nickname and the previous name on the card. So we can generate the display name already when we generate a nickname or take it from the user, and stored in the RDF. In other words, we don't really need to split the nickname for nicer display if that make really a difference.

I am not sure I understand this, but ...

> I will go for the easy way.

Me too. Let's stick to all-dash-nicknames that are not connected to the `name` field, only to the (resource) id.

**Pull request merged**